lib/jekyll/site.rb

```diff
@@ -215,36 +215,45 @@ def posts
       collections["posts"] ||= Collection.new(self, "posts")
     end
 
-    # Construct a Hash of Posts indexed by the specified Post attribute.
+    # Construct a Hash of Documents indexed by the specified Document attribute.
     #
-    # post_attr - The String name of the Post attribute.
+    # doc_attr - The String name of the Document attribute.
+    # documents - The Array of Documents
     #
     # Examples
     #
-    #   post_attr_hash('categories')
-    #   # => { 'tech' => [<Post A>, <Post B>],
-    #   #      'ruby' => [<Post B>] }
+    #   doc_attr_hash('categories')
+    #   # => { 'tech' => [<Document A>, <Document B>],
+    #   #      'ruby' => [<Document B>] }
     #
-    # Returns the Hash: { attr => posts } where
-    #   attr  - One of the values for the requested attribute.
-    #   posts - The Array of Posts with the given attr value.
-    def post_attr_hash(post_attr)
-      # Build a hash map based on the specified post attribute ( post attr =>
-      # array of posts ) then sort each array in reverse order.
+    # Returns the Hash: { attr => docs } where
+    #   attr - One of the values for the requested attribute.
+    #   docs - The Array of Documents with the given attr value.
+    def doc_attr_hash(doc_attr, documents)
+      # Build a hash map based on the specified document attribute
+      # ( doc_attr => array of docs ) then sort each array in reverse order.
       hash = Hash.new { |h, key| h[key] = [] }
-      posts.docs.each do |p|
-        p.data[post_attr].each { |t| hash[t] << p } if p.data[post_attr]
+
+      documents.each do |d|
+        d.data[doc_attr].each { |t| hash[t] << d } if d.data[doc_attr]
       end
-      hash.values.each { |posts| posts.sort!.reverse! }
+
+      hash.values.each { |docs| docs.sort!.reverse! }
       hash
     end
 
+    # As doc_attr_hash, but for posts only
+    # (Provided for backwards compatibility)
+    def post_attr_hash(post_attr)
+      doc_attr_hash(post_attr, posts.docs)
+    end
+
     def tags
-      post_attr_hash("tags")
+      doc_attr_hash("tags", documents)
```

Comment:
> Clearly that is a breaking change, so perhaps something to postpone to a v4.0 release.

Yup.

If `tags` and `categories` were left unchanged, the rest of this would be acceptable.